pkg/plugin/handle_node.go

```diff
@@ -0,0 +1,169 @@
+	newNode, err := state.NodeStateFromK8sObj(node, s.config.Watermark, s.metrics.nodes.inheritedLabels)
+	if err != nil {
+		return fmt.Errorf("could not get state from Node object: %w", err)
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var updated *nodeState
+
+	oldNS, ok := s.nodes[node.Name]
+	if !ok {
+		if expectExists {
+			logger.Warn("Adding node that unexpectedly doesn't exist in local state")
+		}
+
+		s.maxNodeCPU = max(s.maxNodeCPU, newNode.CPU.Total)
+		s.maxNodeMem = max(s.maxNodeMem, newNode.Mem.Total)
+
+		entry := &nodeState{
+			node:                newNode,
+			requestedMigrations: make(map[types.UID]struct{}),
+			podsVMPatchedAt:     make(map[types.UID]time.Time),
+		}
+
+		logger.Info("Adding base node state", zap.Object("Node", entry.node))
+		s.nodes[node.Name] = entry
+		updated = entry
+	} else /* oldNode DOES exist, let's update it */ {
+		if !expectExists {
+			logger.Warn("Updating node that unexpectedly exists in local state")
+		}
+
+		// Use (*Node).Speculatively() so that we can log both states before committing, and provide
+		// protection from panics if .Update() has issues.
+		oldNS.node.Speculatively(func(n *state.Node) (commit bool) {
+			changed := n.Update(newNode)
+			if changed {
+				logger.Warn("Updating base node state", zap.Object("OldNode", oldNS.node), zap.Object("Node", n))
+			}
+			return true // yes, apply the change
+		})
+		updated = oldNS
+	}
+
+	return s.reconcileNode(logger, updated)
+}
+
+func (s *PluginState) deleteNode(logger *zap.Logger, node *corev1.Node, expectExists bool) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	n, exists := s.nodes[node.Name]
+	if exists && !expectExists {
+		logger.Warn("Deleting node that unexpectedly exists in local state")
+	} else if !exists && expectExists {
+		logger.Warn("No-op deleting node that unexpectedly doesn't exist in local state")
+	}
+
+	if exists {
+		s.cleanupNode(logger, n)
+	}
+
+	return nil
+}
+
+// reconcileNode makes any updates necessary given the current state of the node.
+// In particular, this method:
+//
+// 1. Triggers live migration if reserved resources are above the watermark; and
+// 2. Updates the prometheus metrics we expose about the node
+//
+// NOTE: this function expects that the caller has acquired s.mu.
+func (s *PluginState) reconcileNode(logger *zap.Logger, ns *nodeState) error {
+	defer s.metrics.nodes.update(ns.node)
+
+	err := s.balanceNode(logger, ns)
+	if err != nil {
+		return fmt.Errorf("could not trigger live migrations: %w", err)
+	}
+
+	return nil
+}
+
+// updateNodeMetricsAndRequeue updates the node's metrics and puts it back in the reconcile queue.
+//
+// This is typically used to force nodes to stay up-to-date after we update a pod on the node, while
+// helping with fairness between time spent reconciling the pod vs the node.
+func (s *PluginState) updateNodeMetricsAndRequeue(logger *zap.Logger, ns *nodeState) {
+	if err := s.requeueNode(ns.node.Name); err != nil {
```

Comment:
question: Can we have a deadlock here? No concrete example from me, just noticed that this call requires some mutexes to be held:
- caller takes `PluginState.mu`
- we're still holding it inside `updateNodeMetricsAndRequeue()`
- `requeueNode` uses `indexedNodeStore.Store.mutex` during lookup
- `nodeWatchStore.NopUpdate` takes `nodeWatchStore.mutex`
- and calls `queue.Enqueue`

And trying to answer my question here:
- I guess `queue` is fully async and doesn't call any handlers, it just sends notifications to workers. So it shouldn't be possible to deadlock a queue.